src/signet.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2019-2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <signet.h>
+
+#include <consensus/merkle.h>
+#include <consensus/params.h>
+#include <hash.h>
+#include <primitives/block.h>
+#include <primitives/transaction.h>
+#include <script/interpreter.h>
+#include <script/standard.h>        // MANDATORY_SCRIPT_VERIFY_FLAGS
+#include <util/system.h>
+
+int GetWitnessCommitmentIndex(const CBlock& block);
+template<typename T> int GetWitnessCommitmentIndex(const T& tx);
+
+static constexpr int NO_WITNESS_COMMITMENT{-1}; // note: this is copied from validation.h, to avoid a circular dependency issue
+
+// Signet block solution checker
+bool CheckBlockSolution(const CBlock& block, const Consensus::Params& consensusParams)
+{
+    std::vector<uint8_t> signet_data;
+    if (!GetWitnessCommitmentSection(block, SIGNET_HEADER, signet_data)) {
+        return error("CheckBlockSolution: Errors in block (block solution missing)");
+    }
+    SimpleSignatureChecker bsc(GetSignetHash(block));
+    CScript challenge(consensusParams.signet_challenge.begin(), consensusParams.signet_challenge.end());
+    CScript solution = CScript(signet_data.begin(), signet_data.end());
+
+    if (!VerifyScript(solution, challenge, nullptr, MANDATORY_SCRIPT_VERIFY_FLAGS, bsc)) {
+        return error("CheckBlockSolution: Errors in block (block solution invalid)");
+    }
+    return true;
+}
+
+uint256 BlockSignetMerkleRoot(const CBlock& block, bool* mutated = nullptr)
+{
+    std::vector<uint256> leaves;
+    leaves.resize(block.vtx.size());
+    {
+        // find and delete signet signature
+        CMutableTransaction mtx(*block.vtx.at(0));
+        SetWitnessCommitmentSection(mtx, SIGNET_HEADER, std::vector<uint8_t>{});
```

Comment:
in commit 7168395ea63d33075a60369f5778eb635208beba

I think the ignored return value should be documented

```suggestion
        (void)SetWitnessCommitmentSection(mtx, SIGNET_HEADER, std::vector<uint8_t>{});
```